File: VcheckSDKDemoIOS/domain/document_response_models.kt
package com.vcheck.demo.dev.domain

import com.google.gson.annotations.SerializedName

data class DocumentUploadResponse(
    @SerializedName("data")
    val data: DocumentUploadResponseData,
    @SerializedName("error_code")
    var errorCode: Int = 0,
    @SerializedName("message")
    var message: String = ""
)

data class DocumentUploadResponseData(
    @SerializedName("status")
    val status: Int,
    @SerializedName("document")
    val document: Int
)

data class DocumentTypesForCountryResponse(
    @SerializedName("data")
    val data: List<DocTypeData>,
    @SerializedName("error_code")
    var errorCode: Int = 0,
    @SerializedName("message")
    var message: String = ""
)

data class DocTypeData(
    @SerializedName("id")
    val id: Int,
    @SerializedName("country")
    val country: String,
    @SerializedName("category")
    val category: Int,
    @SerializedName("min_pages_count")
    val minPagesCount: Int,
    @SerializedName("max_pages_count")
    val maxPagesCount: Int,
    @SerializedName("auto")
    val auto: Boolean,
    @SerializedName("fields")
    val fields: List<DocField>
)

data class DocField(
    @SerializedName("name")
    val name: String,
    @SerializedName("title")
    val title: DocTitle,
    @SerializedName("type")
    val type: String,
    @SerializedName("regex")
    val regex: String? = null
)

data class DocFieldWitOptPreFilledData(
    val name: String,
    val title: DocTitle,
    val type: String,
    val regex: String?,
    var autoParsedValue: String = ""
)

data class DocTitle(
    @SerializedName("uk")
    val ua: String?,
    @SerializedName("en")
    val en: String,
    @SerializedName("ru")
    val ru: String?
)

// --- PRE-PROCESSED DOC

data class PreProcessedDocumentResponse(
    @SerializedName("data")
    val data: PreProcessedDocData,
    @SerializedName("error_code")
    var errorCode: Int = 0,
    @SerializedName("message")
    var message: String = ""
)

data class PreProcessedDocData(
    @SerializedName("id")
    val id: Int,
    @SerializedName("images")
    val images: ArrayList<String> = arrayListOf(),
    @SerializedName("is_primary")
    val isPrimary: Boolean,
    @SerializedName("parsed_data")
    val parsedData: ParsedDocFieldsData,
    @SerializedName("status")
    val status: Int,
    @SerializedName("type")
    val type: DocTypeData
)

data class ParsedDocFieldsData(
    @SerializedName("date_of_birth")
    var dateOfBirth: String? = null,
    @SerializedName("date_of_expiry")
    var dateOfExpiry: String? = null,
    @SerializedName("name")
    var name: String? = null,
    @SerializedName("number")
    var number: String? = null,
    @SerializedName("og_name")
    var ogName: String? = null,
    @SerializedName("og_surname")
    var ogSurname: String? = null,
    @SerializedName("surname")
    var surname: String? = null
)

File: VcheckSDKDemoIOS/domain/DocumentResponseModels.swift
//
//  DocumentResponseModels.swift
//  VcheckSDKDemoIOS
//
//  Created by Kirill Kaun on 28.04.2022.
//

import Foundation


struct DocumentUploadResponse: Codable {

  var data      : DocumentUploadResponseData? = nil
  var errorCode : Int?    = nil
  var message   : String? = nil

  enum CodingKeys: String, CodingKey {

    case data      = "data"
    case errorCode = "error_code"
    case message   = "message"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    data      = try values.decodeIfPresent(DocumentUploadResponseData.self , forKey: .data      )
    errorCode = try values.decodeIfPresent(Int.self    , forKey: .errorCode )
    message   = try values.decodeIfPresent(String.self , forKey: .message   )
 
  }

  init() {}

}



struct DocumentUploadResponseData: Codable {
    
    var status: Int? = nil
    var document: Int? = nil
    
    enum CodingKeys: String, CodingKey {
        case status = "status"
        case document = "document"
    }
    
    init(from decoder: Decoder) throws {
      let values = try decoder.container(keyedBy: CodingKeys.self)
      status = try values.decodeIfPresent(Int.self, forKey: .status)
      document = try values.decodeIfPresent(Int.self, forKey: .document)
    }
}

// ----- COMMON / FOR COUNTRY

struct DocumentTypesForCountryResponse: Codable {

  var data      : [DocTypeData]? = []
  var errorCode : Int?    = nil
  var message   : String? = nil

  enum CodingKeys: String, CodingKey {
    case data      = "data"
    case errorCode = "error_code"
    case message   = "message"
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    data      = try values.decodeIfPresent([DocTypeData].self , forKey: .data      )
    errorCode = try values.decodeIfPresent(Int.self    , forKey: .errorCode )
    message   = try values.decodeIfPresent(String.self , forKey: .message   )
  }

  init() {}
}


struct DocTypeData: Codable {

  var auto          : Bool?     = nil
  var category      : Int?      = nil
  var country       : String?   = nil
  var fields        : [DocField]? = []
  var id            : Int?      = nil
  var maxPagesCount : Int?      = nil
  var minPagesCount : Int?      = nil

  enum CodingKeys: String, CodingKey {

    case auto          = "auto"
    case category      = "category"
    case country       = "country"
    case fields        = "fields"
    case id            = "id"
    case maxPagesCount = "max_pages_count"
    case minPagesCount = "min_pages_count"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    auto          = try values.decodeIfPresent(Bool.self     , forKey: .auto          )
    category      = try values.decodeIfPresent(Int.self      , forKey: .category      )
    country       = try values.decodeIfPresent(String.self   , forKey: .country       )
    fields        = try values.decodeIfPresent([DocField].self , forKey: .fields        )
    id            = try values.decodeIfPresent(Int.self      , forKey: .id            )
    maxPagesCount = try values.decodeIfPresent(Int.self      , forKey: .maxPagesCount )
    minPagesCount = try values.decodeIfPresent(Int.self      , forKey: .minPagesCount )
 
  }

  init() {}

}



struct DocField: Codable {

  var name  : String? = nil
  var regex : String? = nil
  var title : DocTitle?  = DocTitle()
  var type  : String? = nil

  enum CodingKeys: String, CodingKey {

    case name  = "name"
    case regex = "regex"
    case title = "title"
    case type  = "type"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    name  = try values.decodeIfPresent(String.self , forKey: .name  )
    regex = try values.decodeIfPresent(String.self , forKey: .regex )
    title = try values.decodeIfPresent(DocTitle.self  , forKey: .title )
    type  = try values.decodeIfPresent(String.self , forKey: .type  )
 
  }

  init() {}

}

struct DocTitle: Codable {

  var en : String? = nil
  var ru : String? = nil
  var uk : String? = nil

  enum CodingKeys: String, CodingKey {

    case en = "en"
    case ru = "ru"
    case uk = "uk"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    en = try values.decodeIfPresent(String.self , forKey: .en )
    ru = try values.decodeIfPresent(String.self , forKey: .ru )
    uk = try values.decodeIfPresent(String.self , forKey: .uk )
 
  }

  init() {}

}


// --- PRE-PROCESSED DOC

struct PreProcessedDocumentResponse: Codable {

  var data      : PreProcessedDocData?   = PreProcessedDocData()
  var errorCode : Int?    = nil
  var message   : String? = nil

  enum CodingKeys: String, CodingKey {

    case data      = "data"
    case errorCode = "error_code"
    case message   = "message"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    data      = try values.decodeIfPresent(PreProcessedDocData.self   , forKey: .data      )
    errorCode = try values.decodeIfPresent(Int.self    , forKey: .errorCode )
    message   = try values.decodeIfPresent(String.self , forKey: .message   )
 
  }

  init() {}

}


struct PreProcessedDocData: Codable {

  var id         : Int?        = nil
  var images     : [String]?   = []
  var isPrimary  : Bool?       = nil
  var parsedData : ParsedData? = ParsedData()
  var status     : Int?        = nil
  var type       : DocTypeData? = DocTypeData()

  enum CodingKeys: String, CodingKey {

    case id         = "id"
    case images     = "images"
    case isPrimary  = "is_primary"
    case parsedData = "parsed_data"
    case status     = "status"
    case type       = "type"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    id         = try values.decodeIfPresent(Int.self        , forKey: .id         )
    images     = try values.decodeIfPresent([String].self   , forKey: .images     )
    isPrimary  = try values.decodeIfPresent(Bool.self       , forKey: .isPrimary  )
    parsedData = try values.decodeIfPresent(ParsedData.self , forKey: .parsedData )
    status     = try values.decodeIfPresent(Int.self        , forKey: .status     )
    type       = try values.decodeIfPresent(DocTypeData.self       , forKey: .type       )
 
  }

  init() {}

}


struct ParsedData: Codable {

  var dateOfBirth  : String? = nil
  var dateOfExpiry : String? = nil
  var name         : String? = nil
  var number       : String? = nil
  var ogName       : String? = nil
  var ogSurname    : String? = nil
  var surname      : String? = nil

  enum CodingKeys: String, CodingKey {

    case dateOfBirth  = "date_of_birth"
    case dateOfExpiry = "date_of_expiry"
    case name         = "name"
    case number       = "number"
    case ogName       = "og_name"
    case ogSurname    = "og_surname"
    case surname      = "surname"
  
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)

    dateOfBirth  = try values.decodeIfPresent(String.self , forKey: .dateOfBirth  )
    dateOfExpiry = try values.decodeIfPresent(String.self , forKey: .dateOfExpiry )
    name         = try values.decodeIfPresent(String.self , forKey: .name         )
    number       = try values.decodeIfPresent(String.self , forKey: .number       )
    ogName       = try values.decodeIfPresent(String.self , forKey: .ogName       )
    ogSurname    = try values.decodeIfPresent(String.self , forKey: .ogSurname    )
    surname      = try values.decodeIfPresent(String.self , forKey: .surname      )
 
  }

  init() {}

}


//data class DocFieldWitOptPreFilledData(
//    val name: String,
//    val title: DocTitle,
//    val type: String,
//    val regex: String?,
//    var autoParsedValue: String = ""
//)

